### PR TITLE
[dhctl] Deckhouse registry mirroring

### PR DESCRIPTION
**Pull request by mvasl**
## Description
Added `mirror` subcommand for dhctl, that allows to copy a particular set of Deckhouse releases to a local filesystem and then push that copy to some third-party registry. 

This subcommand does not support copying of Community Edition images.

Example of copying current Deckhouse-release channels images to local filesystem mounted into `/d8-images`:

```bash
dhctl mirror -l "license-token" -i /d8-images
```

Or, to copy Deckhouse images starting from release 1.49 to local filesystem mounted into `/d8-images`.
```bash
dhctl mirror --license="license-token" --min-version="1.49" --images=/d8-images
```

To copy downloaded images into third-party registry:

```bash
dhctl mirror --images="/d8-images" --registry="registry.infra.int:5000" --registry-login="user" --registry-password="pass" 
```

Or in short flags

```bash
dhctl mirror -i "/d8-images" -r "registry.infra.int:5000" -u "user" -p "pass"
```

See `dhctl help mirror` for more information on flags:
```
usage: dhctl mirror --images="PATH" [<flags>]

Copy Deckhouse images from Deckhouse registry to local filesystem and to third-party registry

Flags:
  -h, --help                     Show context-sensitive help (also try --help-long and --help-man).
      --logger-type=pretty       Format logs output of a dhctl in different ways. ($DHCTL_CLI_LOGGER_TYPE)
      --tmp-dir="/var/folders/c0/rprsq8c55kb0c__gchv8fmz80000gn/T/dhctl"  
                                 Set temporary directory for debug purposes. ($DHCTL_CLI_TMP_DIR)
      --version                  Show application version.
  -l, --license=LICENSE          Pull Deckhouse images to local machine using license key. Conflicts with --registry. ($DHCTL_CLI_MIRROR_LICENSE)
  -r, --registry=REGISTRY        Push Deckhouse images to your private registry, specified as registry-host[:port]. Conflicts with --license. ($DHCTL_CLI_MIRROR_PRIVATE_REGISTRY)
  -u, --registry-login=REGISTRY-LOGIN  
                                 Username to log into your registry. ($DHCTL_CLI_MIRROR_USER)
  -p, --registry-password=REGISTRY-PASSWORD  
                                 Password to log into your registry. ($DHCTL_CLI_MIRROR_PASS)
      --fe                       Copy Flant Edition images instead of Enterprise Edition. ($DHCTL_CLI_MIRROR_FE)
  -v, --min-version=MIN-VERSION  Minimal Deckhouse release to copy. Cannot be above current Rock Solid release. ($DHCTL_CLI_MIRROR_MIN_VERSION)
  -i, --images                   Directory for pulled images. ($DHCTL_CLI_MIRROR_IMAGES_PATH)
      --insecure                 Skip TLS checks.

```

## Why do we need it, and what problem does it solve?

This is supposed to replace `d8-pull.sh`/`d8-push.sh` method of installation in air-gapped enviroments

## Checklist
- [ ] The code is covered by unit tests.
- [x] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: dhctl
type: feature
summary: Implemented copying of Deckhouse images to third-party registries for air-gapped installation.
impact_level: default
```

### CODE REVIEW

File: dhctl/cmd/dhctl/commands/mirror.go
Comment (name212) [inline dhctl/cmd/dhctl/commands/mirror.go:1]: we need to build deckhouse miror only ee and ce version 
like here https://github.com/deckhouse/deckhouse/pull/5131/files#diff-0fc921a8a3ad48eda2e4f217e64da1aceb58a432cd125dd933cb257c02b8bbcbR1

File: dhctl/cmd/dhctl/commands/mirror.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+	"gopkg.in/alecthomas/kingpin.v2"
+
+	"github.com/deckhouse/deckhouse/dhctl/pkg/app"
+	"github.com/deckhouse/deckhouse/dhctl/pkg/log"
+	"github.com/deckhouse/deckhouse/dhctl/pkg/operations"
+	"github.com/deckhouse/deckhouse/dhctl/pkg/operations/mirror"
+)
+
+func DefineMirrorCommand(parent *kingpin.Application) *kingpin.CmdClause {
+	cmd := parent.Command("mirror", "Copy Deckhouse registry for air-gaped installation.")
```
Comment (name212) [inline dhctl/cmd/dhctl/commands/mirror.go:31]:
```suggestion
	cmd := parent.Command("mirror", "Copy Deckhouse images from Deckhouse registry to local file system")
```

File: dhctl/cmd/dhctl/commands/mirror.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+	app.DefineMirrorFlags(cmd)
+
+	cmd.Action(func(c *kingpin.ParseContext) error {
+		if app.MirrorRegistryHost != "" {
+			return log.Process("mirror", "Push mirrored Deckhouse Registry images to private registry", func() error {
```
Comment (name212) [inline dhctl/cmd/dhctl/commands/mirror.go:36]:
```suggestion
			return log.Process("mirror", "Push mirrored Deckhouse images from local file system to private registry", func() error {
```

File: dhctl/pkg/operations/mirror.go [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+	"net/url"
+	"os"
+	"path/filepath"
+
+	"github.com/Masterminds/semver/v3"
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/google/go-containerregistry/pkg/v1/layout"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+
+	// "github.com/google/go-containerregistry/pkg/v1"
```
Comment (name212) [inline dhctl/pkg/operations/mirror.go:28]: dead code
```suggestion
```

File: dhctl/pkg/operations/mirror/validate.go [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+import (
+	"fmt"
+
+	"github.com/google/go-containerregistry/pkg/v1/layout"
+	"github.com/google/go-containerregistry/pkg/v1/validate"
+)
+
+type ValidationMode string
+
+const (
```
Comment (name212) [inline dhctl/pkg/operations/mirror/validate.go:26]: duplicated with it. Need to keep only one source of true 
https://github.com/deckhouse/deckhouse/pull/6257/files#diff-0596f6ae70c0bb6f7ee4ecb270c0d20ab22e959a76cf213b4dd0b2f25409d889R34

File: dhctl/pkg/operations/mirror/index.go [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+	"os"
+	"path/filepath"
+	"regexp"
+
+	"github.com/Masterminds/semver/v3"
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/google/go-containerregistry/pkg/v1/layout"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+
+	"github.com/deckhouse/deckhouse/dhctl/pkg/util/maputil"
```
Comment (name212) [inline dhctl/pkg/operations/mirror/index.go:29]: sort imports

File: dhctl/cmd/dhctl/commands/mirror.go [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+	"gopkg.in/alecthomas/kingpin.v2"
+
+	"github.com/deckhouse/deckhouse/dhctl/pkg/app"
+	"github.com/deckhouse/deckhouse/dhctl/pkg/log"
+	"github.com/deckhouse/deckhouse/dhctl/pkg/operations"
+	"github.com/deckhouse/deckhouse/dhctl/pkg/operations/mirror"
+)
+
+func DefineMirrorCommand(parent *kingpin.Application) *kingpin.CmdClause {
+	cmd := parent.Command("mirror", "Copy Deckhouse images from Deckhouse registry to local filesystem")
```
Comment (RomanenkoDenys) [inline dhctl/cmd/dhctl/commands/mirror.go:31]:
```suggestion
	cmd := parent.Command("mirror", "Copy Deckhouse images from Deckhouse registry to local filesystem and back to third-party registry")
```

File: dhctl/cmd/dhctl/commands/mirror.go [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+	app.DefineMirrorFlags(cmd)
+
+	cmd.Action(func(c *kingpin.ParseContext) error {
+		if app.MirrorRegistryHost != "" {
+			return log.Process("mirror", "Push mirrored Deckhouse images from local filesystem to private registry", func() error {
+				return mirrorPushDeckhouseToPrivateRegistry()
+			})
+		}
+
+		return log.Process("mirror", "Pull Deckhouse to local filesystem", func() error {
```
Comment (RomanenkoDenys) [inline dhctl/cmd/dhctl/commands/mirror.go:41]:
```suggestion
		return log.Process("mirror", "Pull Deckhouse images from registry to local filesystem", func() error {
```

File: dhctl/pkg/app/mirror.go [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+		var err error
+
+		if MirrorRegistryHost == "" && MirrorDHLicenseToken == "" {
+			return errors.New("One of --license or --registry is required.")
+		}
+
+		if MirrorRegistryHost != "" && MirrorDHLicenseToken != "" {
+			return errors.New("You have specified both --license and --registry flags. This is not how it works.\n\n" +
+				"Leave only --license if you want to pull Deckhouse from public registry.\n" +
+				"Leave only --registry if you already pulled Deckhouse and want to push it to your private registry.")
```
Comment (RomanenkoDenys) [inline dhctl/pkg/app/mirror.go:104]:
```suggestion
				"Leave only --license if you want to pull Deckhouse images from public registry.\n" +
				"Leave only --registry if you already pulled Deckhouse images and want to push it to your private registry.")
```

File: dhctl/pkg/operations/mirror/index.go [hunk 9]
```diff
@@ -0,0 +1,289 @@
...
+	rootFolder string,
+	modules []Module,
+) (*ImageLayouts, error) {
+	var err error
+	layouts := &ImageLayouts{Modules: map[string]ModuleImageLayout{}}
+
+	fsPaths := map[*layout.Path]string{
+		&layouts.Deckhouse:      filepath.Join(rootFolder, registryRepo),
+		&layouts.Install:        filepath.Join(rootFolder, registryRepo+"/install"),
+		&layouts.ReleaseChannel: filepath.Join(rootFolder, registryRepo+"/release-channel"),
```
Comment (RomanenkoDenys) [inline dhctl/pkg/operations/mirror/index.go:64]:
```suggestion
		&layouts.Install:        filepath.Join(rootFolder, registryRepo, "/install"),
		&layouts.ReleaseChannel: filepath.Join(rootFolder, registryRepo, "/release-channel"),
```

File: dhctl/pkg/operations/mirror/index.go [hunk 10]
```diff
@@ -0,0 +1,289 @@
...
+			return nil, fmt.Errorf("create OCI Image Layout at %s: %w", fsPath, err)
+		}
+		*layoutPtr, err = layout.FromPath(fsPath)
+		if err != nil {
+			return nil, fmt.Errorf("get OCI Image Layout from %s: %w", err)
+		}
+	}
+
+	for _, module := range modules {
+		path := filepath.Join(rootFolder, registryRepo+"/modules/"+module.Name)
```
Comment (RomanenkoDenys) [inline dhctl/pkg/operations/mirror/index.go:77]:
```suggestion
		path := filepath.Join(rootFolder, registryRepo, "modules", module.Name)
```

File: dhctl/pkg/operations/mirror/index.go [hunk 11]
```diff
@@ -0,0 +1,289 @@
...
+		if err := createEmptyImageLayoutAtPath(path); err != nil {
+			return nil, fmt.Errorf("create OCI Image Layout at %s: %w", path, err)
+		}
+		moduleLayout, err := layout.FromPath(path)
+		if err != nil {
+			return nil, fmt.Errorf("get OCI Image Layout from %s: %w", err)
+		}
+
+		path = filepath.Join(rootFolder, registryRepo+"/modules/"+module.Name+"/release")
```
Comment (RomanenkoDenys) [inline dhctl/pkg/operations/mirror/index.go:86]:
```suggestion
		path = filepath.Join(rootFolder, registryRepo, "modules", module.Name, "release")
```

File: dhctl/pkg/operations/mirror/index.go [hunk 12]
```diff
@@ -0,0 +1,289 @@
...
+		Digest    string `json:"digest,omitempty"`
+	} `json:"manifests"`
+}
+
+type ociLayout struct {
+	ImageLayoutVersion string `json:"imageLayoutVersion"`
+}
+
+func FillLayoutsImages(mirrorCtx *Context, layouts *ImageLayouts, deckhouseVersions []*semver.Version) {
+	layouts.DeckhouseImages = map[string]struct{}{
```
Comment (RomanenkoDenys) [inline dhctl/pkg/operations/mirror/index.go:155]: May be cycle be more convenient

File: dhctl/pkg/operations/mirror/modules.go [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+	}
+
+	result := make([]Module, 0, len(modules))
+	for _, module := range modules {
+		m := Module{
+			Name:         module,
+			RegistryPath: fmt.Sprintf("%s/modules/%s", mirrorCtx.RegistryRepo, module),
+			Releases:     []string{},
+		}
+		m.Releases, err = crane.ListTags(m.RegistryPath + "/release")
```
Comment (RomanenkoDenys) [inline dhctl/pkg/operations/mirror/modules.go:54]: why crane ?

File: dhctl/pkg/operations/mirror/versions.go [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+
+	return versionsAboveMinimal, nil
+}
+
+func getTagsFromRegistry(mirrorCtx *Context) ([]string, error) {
+	craneOpts := []crane.Option{}
+	if mirrorCtx.RegistryAuth != nil {
+		craneOpts = append(craneOpts, crane.WithAuth(mirrorCtx.RegistryAuth))
+	}
+	tags, err := crane.ListTags(mirrorCtx.RegistryRepo+"/release-channel", craneOpts...)
```
Comment (RomanenkoDenys) [inline dhctl/pkg/operations/mirror/versions.go:60]: why crane ?

**Review by RomanenkoDenys — CHANGES_REQUESTED** [review]
Some nits

**Comment by mvasl** [comment]
Docker registry
Снимок экрана 2023-10-25 в 17 21 39

**Comment by mvasl** [comment]
Red Hat Quay 3.9.3
Снимок экрана 2023-10-25 в 17 57 51
Снимок экрана 2023-10-25 в 17 57 37
Снимок экрана 2023-10-25 в 17 59 02

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @mvasl. I've started the workflow for label 'e2e/run/openstack'...

**Review by name212 — APPROVED** [review]

**Review by RomanenkoDenys — APPROVED** [review]
LGTM

**Pull request merged**